### PR TITLE
Dockerfile tests and docs

### PR DESCRIPTION
**Pull request by orndorffgrant**
## Proposed Commit Message

```
tests: test using ua in docker builds
```

## Test Steps

Try both of the following:
```
env UACLIENT_BEHAVE_BUILD_PR=1 tox -e behave-docker
```
```
tox -e behave-docker
```

## Desired commit type

 - [ ] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [x] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist

 - [ ] I have updated or added any unit tests accordingly
 - [x] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

File: features/docker.feature
Comment (orndorffgrant) [inline features/docker.feature:30]: I thought this was true at one point during testing but I'll try adding it back in

File: features/docker.feature
Comment (orndorffgrant) [inline features/docker.feature:20]: This line slipped in at some point but isn't necessary, I'll remove it

File: features/docker.feature
Comment (orndorffgrant) [inline features/docker.feature:41]: This was left in from debuggint too, I'll remove

File: features/docker.feature
Comment (orndorffgrant) [inline features/docker.feature:57]: And I need to uncomment these

File: features/docker.feature
Comment (orndorffgrant) [reply features/docker.feature:29]: The comment here is accurate

File: features/steps/steps.py
Comment (orndorffgrant) [reply features/steps/steps.py:481]: @lucasmoura This is a good idea! And it is something I looked into briefly but didn't go too far. Maybe you can help though, because my lack of apt knowledge held me back a bit here.

Is there an easy way to `apt download` a `xenial` package when on a `focal` machine?

I'd rather keep the host machine of this test `focal` and just change the container base series version. The main reason is that `xenial`'s `docker` doesn't support mount secrets, which are necessary for this workflow :/

File: features/docker.feature
Comment (orndorffgrant) [reply features/docker.feature:54]: We can do that too! I just only included the files with secrets in them because they're the most important.

But I think we might as well check for a couple other files too, I'll add them

File: features/docker.feature
Comment (orndorffgrant) [reply features/docker.feature:2]: Oh thats a good idea, we should make sure any `ua` error properly breaks the build. And it should be easy to add once this main test is ironed out. I'll add it

File: features/steps/steps.py
Comment (lucasmoura) [reply features/steps/steps.py:481]: I think we can use `ubuntu-dev-tools` for this. The command `pull-lp-debs ubuntu-advantage-tools xenial` should download the latest version of the package (even if it is on proposed). It also seems that we can specify a ppa to pull the package from, so maybe we can adapt it for the daily case. But it is worth testing it out

Comment (orndorffgrant) [reply features/steps/steps.py:481]: oh very cool! I'll definitely test that out - thanks!

**Comment by lucasmoura** [comment]
@athos-ribeiro could you please take a look at this PR ? I think your docker knowledge will be really useful to see if we are properly testing the docker image we are creating

File: features/docker.feature
Comment (athos-ribeiro) [inline features/docker.feature:33]: is this correct?

File: features/steps/steps.py
Comment (athos-ribeiro) [inline features/steps/steps.py:911]: Since we are already running a shell command here, why not using `'find /var/lib/docker/overlay2 -name {}.format(file_name)'` and checking if anything was pushed to stdout (no need for regular expressions)?

File: features/steps/steps.py
Comment (orndorffgrant) [reply features/steps/steps.py:969]: Thats a great idea! Will do

File: features/steps/steps.py
Comment (orndorffgrant) [reply features/steps/steps.py:911]: yes that is a much better approach, for some reason I never think to use `find` haha, will do

File: features/steps/steps.py
Comment (orndorffgrant) [reply features/steps/steps.py:935]: Ah very good catch! This was there from an earlier iteration where I was running `upgrade` on both the main test image and this internal comparison image. I will re-consider all of this

File: features/docker.feature
Comment (orndorffgrant) [reply features/docker.feature:34]: No but there's no reason to because it is already in a previous layer, so removing won't affect the size. And we're compensating for the deb file in the size comparison test

I suppose we could use a squashing tool like you suggest above, but I don't want to deviate too much from what we expect a customer would do

File: features/docker.feature
Comment (orndorffgrant) [reply features/docker.feature:36]: Because this is happening in the same `RUN` as the `apt update` that downloads the lists in the first place, this does keep the image smaller by about 30Mb in the test I just ran.

**Comment by orndorffgrant** [comment]
@athos-ribeiro I think I've addressed all of your comments, please take another look when you have time :)

File: features/docker.feature
Comment (athos-ribeiro) [reply features/docker.feature:36]: Oh, of course! NVM :)
+1

**Review by athos-ribeiro — APPROVED** [review]
From a containers perspective, this LGTM.

It would be nice to display any contents that gets included in the 2MB margin that was left for comparing image sizes so we could gradually reduce that in the future though (it should be feasible to account for every byte in those layers).

**Comment by orndorffgrant** [comment]
Thanks @athos-ribeiro!

That's an excellent point about the 2MB margin. Do you know of a good way to display the difference in contents? I found this tool https://github.com/GoogleContainerTools/container-diff which looks interesting. Or we could maybe just do something like `find / -type f | sort | tee image-file-list.txt` for each image and diff them (with some extra steps to get sizes). Any suggestions?

**Comment by athos-ribeiro** [comment]
@orndorffgrant: you could get a list with each of the images layers first (you can get that list with `docker inspect` as well). Then you can discard the layers they have in common (they will have at  the base image layers in common), and proceed to check the layer directories to get that file list you want (note that the directories in the overlayfs directories are named after the layer hashes). Then you could even compare files for sizes (or even contents) when the same files are present in different layers and also report on files that are present in one image and not in the other.  You could also pre-fetch the list of files of any packages that are expected to be in one image and not in the other, and add them to some sort of denylist so they won't be flagged by this process.

**Comment by orndorffgrant** [comment]
@athos-ribeiro Thank you! Those are great tips! I've created a Jira issue for that (SC-837) so as to not block this PR.

File: docs/howtoguides/enable_ua_in_dockerfile.md
Comment (orndorffgrant) [inline docs/howtoguides/enable_ua_in_dockerfile.md:1]: Hello @valentincanonical please take a look at the generic how to guide here as well as the fips-centric tutorial below and let us know if you have any feedback :)

Comment (orndorffgrant) [reply docs/howtoguides/enable_ua_in_dockerfile.md:1]: Thanks @valentincanonical!

Totally agree on all points and will make those changes.

Some specific responses:

> why does ca-certificates have to be explicitly installed instead of being an ubuntu-advantage-tools dependency?
We're generally very conservative about adding explicit dependencies because it adds bloat to all ubuntu-minimal-based machines in existence. `ua` will actually partially function without `ca-certificates` and will tell you to install it if it's needed for a given operation. We may want to consider adding this as an actual dependency though (or perhaps a "Recommended" dep), CC @lucasmoura @renanrodrigo @CalvoM @blackboxsw.

> FROM ubuntu:bionic: does it not work with focal?
It does work with focal. I just arbitrarily chose a release for the example, but you're right that we should default to our latest LTS. I'll change it :)

Comment (lucasmoura) [reply docs/howtoguides/enable_ua_in_dockerfile.md:1]: Yes, I agree with @renanrodrigo point here

File: docs/howtoguides/enable_ua_in_dockerfile.md
Comment (lucasmoura) [inline docs/howtoguides/enable_ua_in_dockerfile.md:3]: s/tool take advantage/tool to take advantage

File: docs/howtoguides/enable_ua_in_dockerfile.md
Comment (lucasmoura) [inline docs/howtoguides/enable_ua_in_dockerfile.md:20]: Do we need the `(read: authenticating and authorizing)` part ?

File: docs/howtoguides/enable_ua_in_dockerfile.md
Comment (lucasmoura) [inline docs/howtoguides/enable_ua_in_dockerfile.md:22]: Should we include `cis, usg and cc-eal` or they are too specific that we should not advertise them here ?

File: docs/howtoguides/enable_ua_in_dockerfile.md
Comment (valentincanonical) [reply docs/howtoguides/enable_ua_in_dockerfile.md:1]: Right, because it's the same package that is used in non-container ubuntu, correct? (and the "Recommended" won't help, as we're excluding recommends.) Then that should do it!

Comment (valentincanonical) [reply docs/howtoguides/enable_ua_in_dockerfile.md:1]: @orndorffgrant I tried the tutorial with the daily PPA on focal and bionic today. Works well!
Only thing I would add is: we might want to add a line at the end to advice to squash the layers (as in, they have a ubuntu-focal-fips Dockerfile to build their FIPS-enabled base image, they squash and push it to a **private** registry, and then they can use it FROM whatever else Dockerfile they need)

File: features/steps/steps.py
Comment (orndorffgrant) [reply features/steps/steps.py:180]: Sorry I never responded here. Yes this is just in case for future features since it was easy to include, but isn't being used right now.

File: features/steps/steps.py
Comment (orndorffgrant) [reply features/steps/steps.py:170]: Good idea! I'll add this

File: docs/howtoguides/enable_ua_in_dockerfile.md
Comment (orndorffgrant) [reply docs/howtoguides/enable_ua_in_dockerfile.md:20]: No not really, I'll remove it

File: docs/howtoguides/enable_ua_in_dockerfile.md
Comment (orndorffgrant) [reply docs/howtoguides/enable_ua_in_dockerfile.md:22]: yeah I'm not sure they make sense in a docker container context :shrug:

File: docs/howtoguides/enable_ua_in_dockerfile.md
Comment (orndorffgrant) [reply docs/howtoguides/enable_ua_in_dockerfile.md:1]: @valentincanonical Glad it worked!
Just updated these docs with your suggestions - let me know what you think.

The UA team did discuss more about adding `ca-certificates` as an explicit dependency - we're going to consider adding it as a "Recommends" (which would mean removing the --no-install-recommends from this guide), but it is going to require more discussion and we shouldn't count on that happening before landing this PR.

With regards to squashing: The current tutorials are designed so that you don't need to squash. Because everything happens in one RUN command, everything is already happening in a single layer. So I'm not sure we should recommend it here. I did add a note to the how-to-guide that mentions squashing as an alternative to using one large RUN command though. Let me know if you think this is enough.

File: features/steps/steps.py
Comment (orndorffgrant) [reply features/steps/steps.py:170]: Added!

**Comment by orndorffgrant** [comment]
@lucasmoura Just updated this addressing your last round of comments :)

**Review by lucasmoura — APPROVED** [review]

**Review by renanrodrigo — APPROVED** [review]

**Comment by orndorffgrant** [comment]
I'm going to go ahead and merge this. Feel free to review more later @valentincanonical and we can always iterate on this content.

**Pull request merged**
